std/src/main/scala/algebra/std/list.scala

```diff
@@ -11,15 +12,21 @@ trait ListInstances {
 }
 
 class ListOrder[A](implicit ev: Order[A]) extends Order[List[A]] {
-  def compare(x: List[A], y: List[A]): Int =
-    (x, y) match {
-      case (Nil, Nil) => 0
-      case (Nil, _) => -1
-      case (_, Nil) => 1
-      case (x :: xs, y :: ys) =>
-        val n = ev.compare(x, y)
-        if (n != 0) n else compare(xs, ys)
-    }
+  def compare(xs: List[A], ys: List[A]): Int = {
+    @tailrec def loop(xs: List[A], ys: List[A]): Int =
+      xs match {
+        case Nil =>
```

Comment:
I wrote a benchmark for Lists only here: https://gist.github.com/tixxit/e80586e62c50bffbdfdb

Here are the results:

```
Benchmark                                        Mode  Samples          Score          Error  Units
a.b.TupleMatchBenchmark.emptyListLeft2Match     thrpt       10  293814120.445 ±  8545233.314  ops/s
a.b.TupleMatchBenchmark.emptyListLeftTuple      thrpt       10  300206447.284 ±  6360647.543  ops/s
a.b.TupleMatchBenchmark.emptyListRight2Match    thrpt       10  216331355.650 ±  5895449.200  ops/s
a.b.TupleMatchBenchmark.emptyListRightTuple     thrpt       10  301973062.006 ±  8877747.407  ops/s
a.b.TupleMatchBenchmark.emptyLists2Match        thrpt       10  376145189.382 ± 29439211.344  ops/s
a.b.TupleMatchBenchmark.emptyListsTuple         thrpt       10  135045001.983 ± 10181310.944  ops/s
a.b.TupleMatchBenchmark.nonEmptyLists2Match     thrpt       10  170827405.981 ±  5368140.256  ops/s
a.b.TupleMatchBenchmark.nonEmptyListsTuple      thrpt       10  207933353.408 ±  5254308.264  ops/s
```

The end result here is that matching on a 2-tuple lists is faster everywhere except for the `(Nil, Nil)` case. Worth looking at further. Also, at least in this case with lists, it seems that using a 2-tuple probably isn't a bad strategy.